### PR TITLE
[Sprint 3] Feature Logic Product Detail

### PR DESCRIPTION
**Pull request by tantuan127**
Feature Logic Product Detail
Link redmine : https://dev.framgia.com/redmine/issues/114084
Task : Detail Product
Effect: N/A

### CODE REVIEW

File: .idea/gradle.xml [hunk 1]
```diff
@@ -1,18 +1,18 @@
...
-    <option name="linkedExternalProjectsSettings">
-      <GradleProjectSettings>
-        <option name="distributionType" value="DEFAULT_WRAPPED" />
-        <option name="externalProjectPath" value="$PROJECT_DIR$" />
-        <option name="modules">
-          <set>
-            <option value="$PROJECT_DIR$" />
-            <option value="$PROJECT_DIR$/app" />
-          </set>
+    <component name="GradleSettings">
```
Comment (daolq3012) [inline .idea/gradle.xml]: Please keep old file

File: app/src/main/java/com/framgia/forder/screen/listProduct/ListProductViewModel.java [hunk 2]
```diff
@@ -82,7 +82,7 @@ public void onAddToCart(Product product) {
 
     @Override
     public void onItemRecyclerViewClick(Object item) {
-        mNavigator.goNextChildFragment(R.id.layout_content, ProductDetailFragment.newInstance(),
+        mNavigator.goNextChildFragment(R.id.layout_content, ProductDetailFragment.newInstance(item),
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/listProduct/ListProductViewModel.java]: check 
```java
if (item instanceof Product) {
mNavigator.goNextChildFragment(R.id.layout_content, ProductDetailFragment.newInstance(item)
}
````

File: app/src/main/java/com/framgia/forder/screen/productdetail/ProductDetailFragment.java [hunk 3]
```diff
@@ -2,38 +2,69 @@
...
-    }
-
-    @Override
-    public void onCreate(Bundle savedInstanceState) {
-        super.onCreate(savedInstanceState);
-        mViewModel = new ProductDetailViewModel();
-
-        ProductDetailContract.Presenter presenter = new ProductDetailPresenter(mViewModel);
-        mViewModel.setPresenter(presenter);
+    public static ProductDetailFragment newInstance(Object item) {
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/productdetail/ProductDetailFragment.java]: change to
```java
public static ProductDetailFragment newInstance(Product product) {
...
}
```

File: app/src/main/java/com/framgia/forder/screen/productdetail/ProductDetailFragment.java [hunk 4]
```diff
@@ -2,38 +2,69 @@
...
     @Nullable
     @Override
     public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
             @Nullable Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        Product product = (Product) getArguments().get(EXTRA_PRODUCT);
+        List<Product> products = new ArrayList<>();
+        ProductShopAdapter productAdapter = new ProductShopAdapter(getActivity(), products);
+        Navigator navigator = new Navigator(getParentFragment());
+
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/productdetail/ProductDetailFragment.java]: remove empty line in here

File: app/src/main/java/com/framgia/forder/screen/productdetail/ProductDetailFragment.java [hunk 5]
```diff
@@ -2,38 +2,69 @@
...
+        mViewModel = new ProductDetailViewModel(product, productAdapter, navigator);
+        RealmApi realmApi = new RealmApi();
+        SharedPrefsApi prefsApi = new SharedPrefsImpl(getActivity());
+
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/productdetail/ProductDetailFragment.java]: remove empty line in here

File: app/src/main/java/com/framgia/forder/screen/productdetail/ProductDetailFragment.java [hunk 6]
```diff
@@ -2,38 +2,69 @@
...
     @Override
     public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
             @Nullable Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        Product product = (Product) getArguments().get(EXTRA_PRODUCT);
+        List<Product> products = new ArrayList<>();
+        ProductShopAdapter productAdapter = new ProductShopAdapter(getActivity(), products);
+        Navigator navigator = new Navigator(getParentFragment());
+
+        mViewModel = new ProductDetailViewModel(product, productAdapter, navigator);
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/productdetail/ProductDetailFragment.java]: add empty line in here

File: app/src/main/java/com/framgia/forder/screen/productdetail/ProductDetailFragment.java [hunk 7]
```diff
@@ -2,38 +2,69 @@
...
+        RealmApi realmApi = new RealmApi();
+        SharedPrefsApi prefsApi = new SharedPrefsImpl(getActivity());
+
+        DomainRepository domainRepository =
+                new DomainRepository(new DomainRemoteDataSource(FOrderServiceClient.getInstance()),
+                        new DomainLocalDataSource(prefsApi, new UserLocalDataSource(prefsApi)));
+        ProductRepository productRepository = new ProductRepository(
+                new ProductRemoteDataSource(FOrderServiceClient.getInstance()),
+                new ProductLocalDataSource(realmApi), domainRepository);
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/productdetail/ProductDetailFragment.java]: add empty line in here

File: app/src/main/java/com/framgia/forder/screen/productdetail/ProductDetailPresenter.java [hunk 8]
```diff
@@ -9,13 +21,61 @@
...
+        Subscription subscription =
+                mProductRepository.addToCart(product).subscribe(new Action1<Void>() {
+                    @Override
+                    public void call(Void aVoid) {
+                        mViewModel.onAddToCartSuccess();
+                    }
+                }, new SafetyError() {
+                    @Override
+                    public void onSafetyError(BaseException error) {
+
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/productdetail/ProductDetailPresenter.java]: remove empty line

File: app/src/main/java/com/framgia/forder/screen/productdetail/ProductDetailPresenter.java [hunk 9]
```diff
@@ -9,13 +21,61 @@
...
+                    public void onSafetyError(BaseException error) {
+                        mViewModel.onGetListProductShopError(error);
+                    }
+                });
+        mCompositeSubscription.add(subscription);
     }
 
     @Override
     public void onStart() {
+        mProductRepository.openTransaction();
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/productdetail/ProductDetailPresenter.java]: Please add ``mProductRepository.closeTransaction();`` in method ``onStop()``

File: app/src/main/java/com/framgia/forder/screen/productdetail/ProductDetailViewModel.java [hunk 10]
```diff
@@ -25,4 +50,108 @@ public void onStop() {
...
+
+    @Override
+    public void onAddToCartError(BaseException exception) {
+        // Todo show dialog message
+
+    }
+
+    @Override
+    public void onAddToCartSuccess() {
+        // Todo show dialog message
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/productdetail/ProductDetailViewModel.java]: Please ask Tri to implement in here

File: app/src/main/java/com/framgia/forder/data/source/ProductDataSource.java [hunk 11]
```diff
@@ -27,5 +27,7 @@
 
     interface RemoteDataSource {
         Observable<List<Product>> getListProduct(int domainId);
+
+        Observable<List<Product>> getListProductShop(int shopId, int domainId);
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/data/source/ProductDataSource.java]: change the name of this method to ``getListProductInShop()``

File: app/src/main/java/com/framgia/forder/screen/productdetail/ProductDetailFragment.java (hunk 7 above)
Comment (tantuan127) [inline app/src/main/java/com/framgia/forder/screen/productdetail/ProductDetailFragment.java]: Added

File: app/src/main/java/com/framgia/forder/screen/productdetail/ProductDetailPresenter.java (hunk 9 above)
Comment (tantuan127) [inline app/src/main/java/com/framgia/forder/screen/productdetail/ProductDetailPresenter.java]: Added

File: app/src/main/java/com/framgia/forder/screen/productdetail/ProductDetailViewModel.java (hunk 10 above)
Comment (tantuan127) [inline app/src/main/java/com/framgia/forder/screen/productdetail/ProductDetailViewModel.java]: Added

File: app/src/main/java/com/framgia/forder/screen/productdetail/ProductDetailFragment.java (hunk 7 above)
Comment (tantuan127) [inline app/src/main/java/com/framgia/forder/screen/productdetail/ProductDetailFragment.java]: Added

File: app/src/main/java/com/framgia/forder/screen/productdetail/ProductDetailViewModel.java (hunk 10 above)
Comment (tantuan127) [inline app/src/main/java/com/framgia/forder/screen/productdetail/ProductDetailViewModel.java]: Added

File: .idea/gradle.xml (hunk 1 above)
Comment (tantuan127) [inline .idea/gradle.xml]: fixed

File: app/src/main/java/com/framgia/forder/screen/listProduct/ListProductViewModel.java [hunk 12]
```diff
@@ -69,21 +72,47 @@ public void onAddToCartError(Throwable throwable) {
 
     @Override
     public void onAddToCartSuccess() {
-        // Todo show dialog message
+        new AlertDialog.Builder(mContext).setIcon(R.drawable.ic_check_success)
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/listProduct/ListProductViewModel.java]: replace by "//TODO show dialog in here"

File: app/src/main/java/com/framgia/forder/screen/listProduct/ListProductViewModel.java [hunk 13]
```diff
@@ -69,21 +72,47 @@ public void onAddToCartError(Throwable throwable) {
...
     }
 
     @Override
-    public void onAddToCart(Product product) {
-        if (product == null) {
-            return;
-        }
-        mPresenter.addToCart(product);
+    public void onAddToCart(final Product product) {
+        final ProgressDialog progressDialog = new ProgressDialog(mContext);
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/listProduct/ListProductViewModel.java]: remove all content this method
declare progress in layout and show/hide this progress bar

File: app/src/main/java/com/framgia/forder/screen/listProduct/ListProductViewModel.java [hunk 14]
```diff
@@ -69,21 +72,47 @@ public void onAddToCartError(Throwable throwable) {
...
     public void onItemRecyclerViewClick(Object item) {
-        mNavigator.goNextChildFragment(R.id.layout_content, ProductDetailFragment.newInstance(),
-                true, Navigator.RIGHT_LEFT, "ProductDetailFragment");
+        if (item instanceof Product) {
+            Product product = (Product) item;
+            mNavigator.goNextChildFragment(R.id.layout_content,
+                    ProductDetailFragment.newInstance(product), true, Navigator.RIGHT_LEFT,
+                    "ProductDetailFragment");
+        } else {
+            // Todo show dialog message
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/listProduct/ListProductViewModel.java]: remove ``else`` condition

File: app/src/main/java/com/framgia/forder/screen/productdetail/ProductDetailFragment.java (hunk 7 above)
Comment (tantuan127) [inline app/src/main/java/com/framgia/forder/screen/productdetail/ProductDetailFragment.java]: Added

File: app/src/main/java/com/framgia/forder/screen/listProduct/ListProductViewModel.java [hunk 15]
```diff
@@ -69,7 +65,7 @@ public void onAddToCartError(Throwable throwable) {
 
     @Override
     public void onAddToCartSuccess() {
-        // Todo show dialog message
+        //TODO show dialog in here
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/listProduct/ListProductViewModel.java]: change to ``//TODO update order quantity of product in here``

File: app/src/main/java/com/framgia/forder/screen/productdetail/ProductDetailViewModel.java [hunk 16]
```diff
@@ -25,4 +49,126 @@ public void onStop() {
...
+        Product product = (Product) item;
+        mNavigator.goNextChildFragment(R.id.layout_content,
+                ProductDetailFragment.newInstance(product), true, Navigator.RIGHT_LEFT,
+                "ProductDetailFragment");
+    }
+
+    @Override
+    public void onAddToCart() {
+        mPresenter.addToCart(mProduct);
+        onAddToCartSuccess();
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/productdetail/ProductDetailViewModel.java]: remove this method

**Pull request merged**